storagenode/pieces/readwrite.go

```diff
@@ -79,16 +193,80 @@ func NewReader(blob storage.BlobReader) (*Reader, error) {
 	if err != nil {
 		return nil, Error.Wrap(err)
 	}
+	formatVersion := blob.GetStorageFormatVersion()
+	if formatVersion >= storage.FormatV1 {
+		if size < V1PieceHeaderReservedArea {
+			return nil, Error.New("invalid piece file for storage format version %d: too small for header (%d < %d)", formatVersion, size, V1PieceHeaderReservedArea)
+		}
+		size -= V1PieceHeaderReservedArea
+	}
 
-	reader := &Reader{}
-	reader.blob = blob
-	reader.size = size
-
+	reader := &Reader{
+		formatVersion: formatVersion,
+		blob:          blob,
+		size:          size,
+	}
 	return reader, nil
 }
 
+// GetStorageFormatVersion returns the storage format version of the piece being read.
+func (r *Reader) GetStorageFormatVersion() storage.FormatVersion {
+	return r.formatVersion
+}
+
+// GetPieceHeader reads, unmarshals, and returns the piece header. It may only be called
+// before any Read() calls. (Retrieving the header at any time could be supported, but for
```

Comment:
We should also mention that this method can only be called once, unless you manually seek back to the start.